assemble/Particle_Diagnostics.F90

```diff
@@ -0,0 +1,1420 @@
+
+          !Loop over particle arrays
+          do i = 1,size(group_arrays)
+             particle => particle_lists(group_arrays(i))%first
+             if (associated(particle)) then !Only work on arrays if local_particles exist on this processor
+                allocate(local_crds(size(particle%local_coords)))
+             end if
+             do while(associated(particle))
+                !Get element, local_crds and attribute value of each particle
+                element = particle%element
+                local_crds = particle%local_coords
+                att_value = particle%attributes(group_attribute)
+
+                !Find nodes for specified element
+                nodes => ele_nodes(s_field, element)
+
+                !Distribute particle values across nodes of element by weighted distance function
+                do j = 1, size(nodes)
+                   node_number = nodes(j)
+                   node_values(node_number) = node_values(node_number) + att_value*local_crds(j)
+                   node_part_count(node_number) = node_part_count(node_number) + 1.0*local_crds(j)
+                end do
+                particle => particle%next
+
+             end do
+             if (allocated(local_crds)) then
+                deallocate(local_crds)
+             end if
+          end do
+
+       else if (.not. is_ratio .or. have_option(trim(complete_field_path(s_field%option_path))// "/algorithm/interpolation/nearest_neighbour")) then
+
+          !Loop over particle arrays
+          do i = 1,size(group_arrays)
+             particle => particle_lists(group_arrays(i))%first
+             if (associated(particle)) then !Only work on arrays if local_particles exist on this processor
+                allocate(local_crds(size(particle%local_coords)))
+             end if
+             do while(associated(particle))
+                !Get element, local_crds and attribute value of each particle
+                element = particle%element
+                local_crds = particle%local_coords
+                if (is_ratio) then
+                   att_value = particle%attributes(group_attribute)
+                end if
+                !Find nodes for specified element
+                nodes => ele_nodes(s_field, element)
+                ! work out nearest node based on max. local coordinate
+                node_number = nodes(maxloc(local_crds, dim=1))
+                if (is_ratio) then
+                   !Store particle attribute value on closest node
+                   node_values(node_number) = node_values(node_number) + att_value
+                end if
+                !Increase particle count for this node by 1
+                node_part_count(node_number) = node_part_count(node_number) + 1.0
+                particle => particle%next
+
+             end do
+             if (allocated(local_crds)) then
+                deallocate(local_crds)
+             end if
+          end do
+       end if
```

Comment:
Does this `if-block` need an else? It's a little unclear what the cases are; Maybe add a comment?